mlco2/codecarbon
# API backend authentication configuration

Draft configuration of authentification router : 
- Based on a Keycloak authentication server 
- Protects API routers, exposes a /login endoint to obtain an access token if credentials are valid & declared in the authorisation server
- Lets user declare a custom client, to publish data from the package without revealing the applicative password.


To test locally : 

- Spawn a local keyclaok instance using the docker image : 
`docker run -p 8080:8080 -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=admin quay.io/keycloak/keycloak:15.0.1
`
- import from keycloak local admin [console](http://localhost:8080/auth/admin/master/console/) the realm configuration
![image](https://user-images.githubusercontent.com/43726203/130354282-e7a0034c-f6a1-49d1-b34a-6e816ce72196.png)

- Import the backend client & generate a secret key

![image](https://user-images.githubusercontent.com/43726203/129012026-0fcf85b8-be15-4bbe-8b4c-e03ec543e3af.png)

- Change secret key & initial access token values you can generate in the freshly imported client :  
![image](https://user-images.githubusercontent.com/43726203/130354521-badfcce6-af10-4bf7-907d-1ae28a5bb945.png)

- Create a test account in the Users tab
![image](https://user-images.githubusercontent.com/43726203/129012099-6d501767-9223-4e84-aee8-be90b32b1454.png)

- Create an access token via the Swagger UI using the test account credentials 
![image](https://user-images.githubusercontent.com/43726203/129011536-ded9a0b8-9731-4890-9537-f6d0dbefa946.png)

- Use the access token to authenticate 
![image](https://user-images.githubusercontent.com/43726203/129011596-7bf7b610-ce4d-4baa-b8eb-c7e6f568d9e0.png)

- Register a new client dynamically. This is currently working from a curl request, the /register_client router should be able to perform the request too (wip).
```bash
curl -X "POST" "http://localhost:8080/auth/realms/master/clients-registrations/default/" \
     -H 'Authorization: access_token' \
     -H 'Content-Type: application/json; charset=utf-8' \
     -d $'{
  "secret": "secret_key",
  "consentRequired": "false",
  "clientId": "prod_example@user.org",
  "serviceAccountsEnabled": "true",
  "directAccessGrantsEnabled": "true"
}'
```


- This new custom client can be used to generate tokens without using the user's credentials, by referencing the client_id / client_secret in a package login method : 
![image](https://user-images.githubusercontent.com/43726203/130354577-24f7faba-88cb-413f-a5d4-eece4428e0e8.png)
